packages/by-name/kata/kata-runtime/0020-genpolicy-support-dynamic-annotations.patch

```diff
@@ -0,0 +1,177 @@
+ 
+     every i_key in i_keys {
+         allow_anno_key(i_key, p_oci)
++        allow_dynamic_anno_value(i_key, i_annotations, p_oci)
+     }
+ 
+     print("allow_anno 2: true")
+@@ -227,6 +231,49 @@ allow_anno_key(i_key, p_oci) {
+     print("allow_anno_key 2: true")
+ }
+ 
++allow_anno_key(i_key, p_oci) {
++    print("allow_anno_key 3: i key =", i_key)
++
++    some p_key, _ in p_oci.DynamicAnnotations
++    regex.match(p_key, i_key)
++
++    print("allow_anno_key 3: true")
++}
++
++
++# Account for containers without DynamicAnnotations
++# at all..
++allow_dynamic_anno_value(i_key, i_annotations, p_oci) {
++    print("allow_dynamic_anno_value 1: i key =", i_key)
++
++    not p_oci.DynamicAnnotations
++
++    print("allow_dynamic_anno_value 1: true")
++}
++# ..for annotations which do not have a corresponding
++# DynamicAnnotation set in the settings...
++allow_dynamic_anno_value(i_key, i_annotations, p_oci) {
++    print("allow_dynamic_anno_value 2: i key =", i_key)
++
++    every p_key, _ in p_oci.DynamicAnnotations {
++      not regex.match(p_key, i_key)
++    }
++
++    print("allow_dynamic_anno_value 2: true")
++}
++# ...and check those which do.
++allow_dynamic_anno_value(i_key, i_annotations, p_oci) {
++    print("allow_dynamic_anno_value 3: i key =", i_key)
++
++    some p_key, p_value in p_oci.DynamicAnnotations
++    regex.match(p_key, i_key)
++
++    i_value := i_annotations[i_key]
++    regex.match(p_value, i_value)
++
++    print("allow_dynamic_anno_value 3: true")
++}
++
+ # Get the value of the S_NAME_KEY annotation and
+ # correlate it with other annotations and process fields.
+ allow_by_anno(p_container, i_oci, i_storages) {
+@@ -292,7 +339,7 @@ allow_by_container_types(p_oci, i_oci, s_name, s_namespace) {
+     print("allow_by_container_types: checking io.kubernetes.cri.container-type")
+ 
+     c_type := "io.kubernetes.cri.container-type"
+-    
++
+     p_cri_type := p_oci.Annotations[c_type]
+     i_cri_type := i_oci.Annotations[c_type]
+     print("allow_by_container_types: p_cri_type =", p_cri_type, "i_cri_type =", i_cri_type)
+diff --git a/src/tools/genpolicy/src/policy.rs b/src/tools/genpolicy/src/policy.rs
+index e2012bf6f2d80ffea678a38803d8e85f5369b9dc..981841eb2680bc60388953a5fbcd4809218f04ac 100644
+--- a/src/tools/genpolicy/src/policy.rs
++++ b/src/tools/genpolicy/src/policy.rs
+@@ -96,6 +96,13 @@ pub struct KataSpec {
+     #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
+     pub Annotations: BTreeMap<String, String>,
+ 
++    /// DynamicAnnotations contains arbitrary metadata for the container.
++    /// It is different to `Annotations` in that it allows dynamic keys *and*
++    /// values, and that they are checked for *all* keys, whereas `Annotations`
++    /// only allows dynamic values, and only checks them for certain keys at all.
++    #[serde(default, skip_serializing_if = "Option::is_none")]
++    pub DynamicAnnotations: Option<BTreeMap<String, String>>,
```

Comment:
The `KataSpec` struct is in fact an OCI container spec. Can we pack this into `PolicyData.common` instead?

Reply:
Do you see it fit there? I figured that this would be a container-scoped setting, and thus put it there. Otherwise we'd also allow it for the pause container.

Reply:
But these are not sourced from the individual containers, but copied from the settings file - which is what the `common` field is for.

Reply:
I meant the `pause_container` / `other_container` sections in `genpolicy-settings.json`. But I'm fine with moving it to `common`, if you see it fit better.

Reply:
This is not yet resolved, is it?

Reply:
It's in `common` now. Unless I misunderstand you, I think that should resolve it?

Reply:
Sorry, I misread the diff.